_posts/2016-01-22-monitoring-an-external-tcp-service-in-elixir.md

```diff
@@ -0,0 +1,378 @@
+  end
+end
+```
+
+When the connection closes, we try to reconnect, once again setting the failure counts to the right values.
+
+# Step 6 callbacks
+
+We're now tracking the complete lifecycle of our tcp connection, so we can focus on exposing callbacks to act on disconnect/reconnect events. There are different strategies we can follow for this: one option is to initialized the `HSC` worker with the `pid` of another process that will receive messages for the aforementioned events, another is to simply pass the callback functions with the rest of the configuration. We'll stick with the latter for now, as the former requires a more extended process infrastructure.
+
+We can revise our application entry point as follows:
+
+```elixir
+  def start(_type, _args) do
+    import Supervisor.Spec, warn: false
+
+    children = [
+      # Define workers and child supervisors to be supervised
+      worker(HSC, [hsc_config]),
+    ]
+    ...
+  end
+  
+  defp hsc_config do
+    [host: "localhost",
+     port: 5432,
+     on_connect: fn(state) -> Logger.info "connected" end,
+     on_disconnect: fn(state) -> Logger.error "disconnected" end]
+  end
+```
+
+Our ideal api exposes two functions, `on_connect/1` and `on_disconnect/1`, that will receive the `HSC` worker state as an argument. This way we can use the state information to print a logline, etc. In the `on_connect/1` function we can do whatever's needed to restore the health of our application, for example calling `Applicaton.ensure_started/2` to restart (if needed) our external service dependant application. If we were monitoring a Postgresql server and using [Ecto](https://github.com/elixir-lang/ecto), we could call:
+
+```elixir
+Application.ensure_started(:poolboy)
+Application.ensure_started(:ecto)
+```
+
+As for the implementation of the two callbacks, we can revise the `HSC` module by extending its `State` struct definition and calling the relevant callbacks where needed:
+
+```elixir
+defmodule HSC do
+  use GenServer
+
+  @max_retries 10
+  @retry_interval 1000
+
+  defmodule DefaultCallbacks do
+    require Logger
+
+    def on_connect(state) do
+      Logger.info("tcp connect to #{state.host}:#{state.port}")
+    end
+
+    def on_disconnect(state) do
+      Logger.info("tcp disconnect from #{state.host}:#{state.port}")
+    end
+  end
+
+  defmodule State do
+    defstruct host: "localhost",
+              port: 1234,
+              failure_count: 0,
+              on_connect: &DefaultCallbacks.on_connect/1,
+              on_disconnect: &DefaultCallbacks.on_disconnect/1
+
+  end
+
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  def init(opts) do
+    state = opts_to_initial_state(opts)
+    case :gen_tcp.connect(state.host, state.port, []) do
+      {:ok, _socket} ->
+        state.on_connect.(state)
+        {:ok, state}
+      {:error, _reason} ->
+        new_state = %{state | failure_count: 1}
+        new_state.on_disconnect.(new_state)
```

Comment:
Technically, this is not disconnect, because you didn't connect in the first place :-)